swupd/packs.go

```diff
@@ -1,10 +1,296 @@
+		}
+
+		done[f.Hash] = true
+
+		// TODO: Pack deltas when available.
+		entry.State = PackedFullfile
+		entry.Reason = "from fullfile"
+		if chrootDir != "" {
+			err, fallback := copyFromChrootFile(tw, chrootDir, f)
+			if (err != nil) && fallback {
+				info.Warnings = append(info.Warnings, err.Error())
+				err = copyFromFullfile(tw, stateDir, f)
+			} else {
+				entry.Reason = "from chroot"
+			}
+		} else {
+			err = copyFromFullfile(tw, stateDir, f)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = tw.Close()
+	if err != nil {
+		return nil, err
+	}
+	return info, nil
+}
+
+func copyFromChrootFile(tw *tar.Writer, chrootDir string, f *File) (err error, fallback bool) {
+	realname := filepath.Join(chrootDir, f.Name)
+	fi, err := os.Lstat(realname)
+	if err != nil {
+		return err, true
+	}
+	hdr, err := getHeaderFromFileInfo(fi)
+	if err != nil {
+		return err, true
+	}
+	hdr.Name = f.Hash.String()
+
+	// TODO: Also perform this verification for copyFromFullfile?
+
+	switch f.Type {
+	case typeDirectory:
+		if !fi.IsDir() {
+			return fmt.Errorf("couldn't use %s for packing: manifest expected a directory but it is not", realname), true
+		}
+		hdr.Typeflag = tar.TypeDir
+	case typeLink:
+		if fi.Mode()&os.ModeSymlink == 0 {
+			return fmt.Errorf("couldn't use %s for packing: manifest expected a link but it is not", realname), true
+		}
+		link, err := os.Readlink(realname)
+		if err != nil {
+			return fmt.Errorf("couldn't use %s for packing: %s", realname, err), true
+		}
+		hdr.Typeflag = tar.TypeSymlink
+		hdr.Linkname = link
+	case typeFile:
+		if !fi.Mode().IsRegular() {
+			return fmt.Errorf("couldn't use %s for packing: manifest expected a regular file but it is not", realname), true
+		}
+		hdr.Typeflag = tar.TypeReg
+	default:
+		return fmt.Errorf("unsupported file %s in chroot with type %q", f.Name, f.Type), true
+	}
+
+	// After we start writing on the tar writer, we can't let the caller fallback to another
+	// option anymore.
+
+	err = tw.WriteHeader(hdr)
+	if err != nil {
+		return err, false
+	}
+
+	if hdr.Typeflag == tar.TypeReg {
+		realfile, err := os.Open(realname)
+		if err != nil {
+			return err, false
+		}
+		defer realfile.Close()
+
+		_, err = io.Copy(tw, realfile)
+		if err != nil {
+			return err, false
+		}
+	}
+	return nil, false
+}
+
+type compressedTarReader struct {
+	*tar.Reader
+	compressionCloser io.Closer
+}
+
+func (ctr *compressedTarReader) Close() error {
+	if ctr.compressionCloser != nil {
+		return ctr.compressionCloser.Close()
+	}
+	return nil
+}
+
+// newCompressedTarReader creates a struct compatible with tar.Reader reading from uncompressed or
+// compressed input.
+func newCompressedTarReader(rs io.ReadSeeker) (*compressedTarReader, error) {
+	var h [6]byte
+	_, err := rs.Read(h[:])
+	if err != nil {
+		return nil, err
+	}
+	_, err = rs.Seek(0, io.SeekStart)
+	if err != nil {
+		return nil, err
+	}
+	result := &compressedTarReader{}
+	switch {
+	case h[0] == 0x1F && h[1] == 0x8B:
```

Comment:
maybe a comment describing what these bytes represent? You are just reading header information to determine what to uncompress with, right?

Reply:
Correct. And I see now how it might be confusing. Added a comment and changed the code to make the intention more explicit.